java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -235,15 +245,33 @@ private boolean shouldTerminate(Set<LearnedConstraint> learnedConstraints) {
       return flowBuilder.build();
     }
 
-    @Nullable
-    private Symbol newTrackedSymbol(ExplodedGraph.Edge edge) {
-      if (trackSymbol == null) {
-        return null;
-      }
-      if (learnedAssociation(edge).isPresent()) {
-        return edge.parent.programState.getLastEvaluated();
-      }
-      return trackSymbol;
+    private PSet<Symbol> newTrackedSymbols(ExplodedGraph.Edge edge) {
+      Optional<LearnedAssociation> learnedAssociation = learnedAssociation(edge);
+      return learnedAssociation.map(la -> {
+        PSet<Symbol> newTrackedSymbols = trackedSymbols.remove(la.symbol);
+        // if assigning literal, stop tracking symbols, because same SV is reused
+        if (isLiteralAssignment(edge.parent.programPoint.syntaxTree())) {
+          return newTrackedSymbols;
+        }
+        ProgramState programState = edge.parent.programState;
+        Symbol symbol = programState.lastAssociatedSymbol.get(la.symbolicValue());
+        if (symbol != null) {
+          newTrackedSymbols = newTrackedSymbols.add(symbol);
+        } else {
+          // take all symbolic values used in this assignment and retrieve corresponding symbols
+          for (SymbolicValue sv : computedFrom(la.symbolicValue())) {
+            Symbol newSymbol = programState.lastAssociatedSymbol.get(sv);
+            if (newSymbol != null) {
+              newTrackedSymbols = newTrackedSymbols.add(newSymbol);
+            }
+          }
+        }
+        return newTrackedSymbols;
+      }).orElse(trackedSymbols);
+    }
+
+    private boolean isLiteralAssignment(Tree tree) {
+      return tree.is(Tree.Kind.ASSIGNMENT) && ((AssignmentExpressionTree) tree).expression().is(Tree.Kind.NULL_LITERAL, Tree.Kind.BOOLEAN_LITERAL);
```

Comment:
This is needed because literals reuse same symbolic value. This means that if we lookup what was last assigned symbol from program state, it can give us some completely unrelated symbol which happens to hold `null`, `true` or `false`. 

Another option would be to create new symbolic value for every literal, we have a ticket here https://jira.sonarsource.com/browse/SONARJAVA-2164